src/dbnode/storage/shard.go

```diff
@@ -2660,74 +2661,217 @@ func (s *dbShard) Repair(
+					if lastIdx < len(frame.Units().Values()) {
+						unit = frame.Units().Values()[lastIdx]
+					}
+				}
+
+				if !singleAnnotation {
+					// TODO: what happens if annotation has changed mid-tile?
+					// Write early and then do the remaining values separately?
+					if lastIdx < len(frame.Annotations().Values()) {
+						annotation = frame.Annotations().Values()[lastIdx]
+					}
+				}
+
+				dp.Timestamp = timestamp
+				dp.Value = value
+				err = encoder.Encode(dp, unit, annotation)
+				if err != nil {
+					s.metrics.largeTilesWriteErrors.Inc(1)
+					multiErr = multiErr.Add(err)
+				}
+
+				processedTileCount.Inc()
+			}
 		}
-		if err != nil {
-			return processedBlockCount, err
+
+		segment := encoder.DiscardReset(opts.Start, segmentCapacity, targetSchemaDescr)
+
+		segmentLen := segment.Len()
+		if segmentLen > segmentCapacity {
+			// Will use the same capacity for the next series.
+			segmentCapacity = segmentLen
 		}
 
-		data.IncRef()
-		bytesReader.Reset(data.Bytes())
-		dataPointIter.Reset(bytesReader, nil)
+		writerData[0] = segment.Head.Bytes()
+		writerData[1] = segment.Tail.Bytes()
+		checksum := segment.CalculateChecksum()
 
-		for dataPointIter.Next() {
-			dp, unit, annot := dataPointIter.Current()
-			_, err = s.writeAndIndex(ctx, id, tags, dp.Timestamp, dp.Value, unit, annot, wOpts, true)
-			if err != nil {
-				s.metrics.largeTilesWriteErrors.Inc(1)
-				lastWriteError = err
-			}
+		if err := writer.WriteAll(id.Bytes(), encodedTags, writerData, checksum); err != nil {
+			s.metrics.largeTilesWriteErrors.Inc(1)
+			multiErr = multiErr.Add(err)
+		} else {
+			s.metrics.largeTilesWrites.Inc(1)
 		}
 
-		dataPointIter.Close()
+		id.Finalize()
+		//FIXME apparently this does not finalize segment.Tail as FinalizeTail flag is not set by default.
+		segment.Finalize()
+	}
 
-		data.DecRef()
-		data.Finalize()
+	if err := writer.Close(); err != nil {
+		multiErr = multiErr.Add(err)
+	} else {
+		// Notify all block leasers that a new volume for the namespace/shard/blockstart
+		// has been created. This will block until all leasers have relinquished their
+		// leases.
+		_, err = s.opts.BlockLeaseManager().UpdateOpenLeases(block.LeaseDescriptor{
+			Namespace:  s.namespace.ID(),
+			Shard:      s.ID(),
+			BlockStart: opts.Start,
+		}, block.LeaseState{Volume: nextVersion})
+	}
 
-		processedBlockCount++
+	if err := readerIter.Err(); err != nil {
+		multiErr = multiErr.Add(err)
+	}
+
+	closed = true
+	if err := readerIter.Close(); err != nil {
+		multiErr = multiErr.Add(err)
+	}
+
+	// If there were some errors shard is still flushable. Log an error but do not return it to the upper level.
+	writeErr := multiErr.FinalError()
+	if writeErr != nil {
+		s.logger.Error("error writing large tiles",
+			zap.Uint32("shardID", sourceShardID),
+			zap.String("sourceNs", sourceNsID.String()),
+			zap.String("targetNs", s.namespace.ID().String()),
+			zap.Error(writeErr),
+		)
```

Comment:
I would actually not log error from this level, I would do it from the calling method, and also return an error from this method (i.e. do not mask the error here).

The top level already adds it to a multiErr and can do what it wants to do with per shard failures, so better not to hide/mask the error to the layer above. The layer above (namespace) can do whatever it wants to do with that error information.

Reply:
👍